Mathlib/Analysis/Calculus/ImplicitContDiff.lean

```diff
@@ -0,0 +1,141 @@
+  one_le : 1 ≤ n
+
+namespace IsContDiffImplicitAt
+
+variable
+  {𝕜 : Type*} [RCLike 𝕜]
+  {E : Type*} [NormedAddCommGroup E] [NormedSpace 𝕜 E] [CompleteSpace E]
+  {F : Type*} [NormedAddCommGroup F] [NormedSpace 𝕜 F] [CompleteSpace F]
+  {n : WithTop ℕ∞} {f : E × F → F} {f' : E × F →L[𝕜] F} {a : E × F}
+
+/-- We record the parameters of our specific case in order to apply the general implicit function
+theorem. -/
+def implicitFunctionData (h : IsContDiffImplicitAt n f f' a) :
+    ImplicitFunctionData 𝕜 (E × F) E F where
+  leftFun := Prod.fst
+  leftDeriv := ContinuousLinearMap.fst 𝕜 E F
+  rightFun := f
+  rightDeriv := f'
+  pt := a
+  hasStrictFDerivAt_leftFun := by fun_prop
+  hasStrictFDerivAt_rightFun := h.contDiffAt.hasStrictFDerivAt' h.hasFDerivAt h.one_le
+  range_leftDeriv := LinearMap.range_eq_top_of_surjective _ fun x ↦ ⟨(x, 0), rfl⟩
+  range_rightDeriv := h.range_eq_top
+  isCompl_ker := by
+    have : ker (ContinuousLinearMap.fst 𝕜 E F) = ker (LinearMap.fst 𝕜 E F) := rfl
+    rw [isCompl_comm, this, LinearMap.ker_fst, h.ker_eq_left]
+    exact LinearMap.isCompl_range_inl_inr
+
+@[simp]
+lemma implicitFunctionData_prodFun (h : IsContDiffImplicitAt n f f' a) :
+    (h.implicitFunctionData.leftFun h.implicitFunctionData.pt,
+      h.implicitFunctionData.rightFun h.implicitFunctionData.pt) = (a.1, f a) := rfl
+
+/-- The implicit function provided by the general theorem, from which we construct the more useful
+form `IsContDiffImplicitAt.implicitFunction`. -/
+noncomputable def implicitFunctionAux (h : IsContDiffImplicitAt n f f' a) : E → F → E × F :=
+  h.implicitFunctionData.implicitFunction
+
+lemma implicitFunctionAux_fst (h : IsContDiffImplicitAt n f f' a) :
+    ∀ᶠ p in 𝓝 (a.1, f a), (h.implicitFunctionAux p.1 p.2).1 = p.1 :=
+  h.implicitFunctionData.prod_map_implicitFunction.mono fun _ ↦ congr_arg Prod.fst
+
+lemma rightFun_implicitFunctionAux (h : IsContDiffImplicitAt n f f' a) :
+    ∀ᶠ p in 𝓝 (a.1, f a), f (h.implicitFunctionAux p.1 p.2) = p.2 :=
+  h.implicitFunctionData.prod_map_implicitFunction.mono fun _ ↦ congr_arg Prod.snd
+
+/-- Implicit function `φ` defined by `f (x, φ x) = f a`. -/
+noncomputable def implicitFunction (h : IsContDiffImplicitAt n f f' a) : E → F :=
+  fun x ↦ (h.implicitFunctionAux x (f a)).2
+
+lemma implicitFunction_def (h : IsContDiffImplicitAt n f f' a) :
+    h.implicitFunction = fun x ↦ (h.implicitFunctionData.implicitFunction.uncurry (x, (f a))).2 :=
+  rfl
+
+/-- `implicitFunction` is indeed the (local) implicit function defined by `f`. -/
+lemma apply_implicitFunction (h : IsContDiffImplicitAt n f f' a) :
+    ∀ᶠ x in 𝓝 a.1, f (x, h.implicitFunction x) = f a := by
+  have := h.rightFun_implicitFunctionAux
+  have hfst := h.implicitFunctionAux_fst
+  rw [nhds_prod_eq, eventually_swap_iff] at this hfst
+  apply this.curry.self_of_nhds.mp
+  apply hfst.curry.self_of_nhds.mono
+  intro x
+  simp_rw [Prod.swap_prod_mk]
+  intro h1 h2
+  rw [← h2]
+  congr 1
+  ext
```

Comment:
Sadly `grind` cannot seem to do the last `rfl`. *If* you can write that as `exact foo` for some lemma `foo`, then `grind [foo]` could have a chance of doing it.